// @mysten/zksend/dist/esm/experimental/wallet/channel/index.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var _popup, _version, _id, _origin, _name, _chain, _promise, _resolve, _reject, _interval, _listener, _StashedPopup_instances, cleanup_fn, _request;
import { parse, safeParse } from "valibot";
import { withResolvers } from "../../utils/withResolvers.js";
import { StashedRequest, StashedResponse } from "./events.js";
const DEFAULT_STASHED_ORIGIN = "https://getstashed.com";
const getClientMetadata = () => {
  return {
    version: "v1",
    originUrl: window.location.href,
    userAgent: navigator.userAgent,
    screenResolution: `${window.screen.width}x${window.screen.height}`,
    language: navigator.language,
    platform: navigator.platform,
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    timestamp: Date.now()
  };
};
class StashedPopup {
  constructor({
    name,
    origin = DEFAULT_STASHED_ORIGIN,
    chain
  }) {
    __privateAdd(this, _StashedPopup_instances);
    __privateAdd(this, _popup);
    __privateAdd(this, _version);
    __privateAdd(this, _id);
    __privateAdd(this, _origin);
    __privateAdd(this, _name);
    __privateAdd(this, _chain);
    __privateAdd(this, _promise);
    __privateAdd(this, _resolve);
    __privateAdd(this, _reject);
    __privateAdd(this, _interval, null);
    __privateAdd(this, _listener, (event) => {
      if (event.origin !== __privateGet(this, _origin)) {
        return;
      }
      const { success, output } = safeParse(StashedResponse, event.data);
      if (!success || output.id !== __privateGet(this, _id)) return;
      __privateMethod(this, _StashedPopup_instances, cleanup_fn).call(this);
      if (output.payload.type === "reject") {
        __privateGet(this, _reject).call(this, new Error("User rejected the request"));
      } else if (output.payload.type === "resolve") {
        __privateGet(this, _resolve).call(this, output.payload.data);
      }
    });
    const popup = window.open("about:blank", "_blank");
    if (!popup) {
      throw new Error("Failed to open new window");
    }
    __privateSet(this, _popup, popup);
    __privateSet(this, _id, crypto.randomUUID());
    __privateSet(this, _origin, origin);
    __privateSet(this, _name, name);
    __privateSet(this, _version, "v1");
    __privateSet(this, _chain, chain);
    const { promise, resolve, reject } = withResolvers();
    __privateSet(this, _promise, promise);
    __privateSet(this, _resolve, resolve);
    __privateSet(this, _reject, reject);
    __privateSet(this, _interval, setInterval(() => {
      try {
        if (__privateGet(this, _popup).closed) {
          __privateMethod(this, _StashedPopup_instances, cleanup_fn).call(this);
          reject(new Error("User closed the Stashed window"));
        }
      } catch {
      }
    }, 1e3));
  }
  send({
    type,
    ...data
  }) {
    window.addEventListener("message", __privateGet(this, _listener));
    const requestData = {
      version: __privateGet(this, _version),
      requestId: __privateGet(this, _id),
      appUrl: window.location.href.split("#")[0],
      appName: __privateGet(this, _name),
      payload: {
        type,
        chain: __privateGet(this, _chain),
        ...data
      },
      metadata: getClientMetadata()
    };
    const encodedRequestData = btoa(JSON.stringify(requestData));
    __privateGet(this, _popup).location.assign(
      `${__privateGet(this, _origin)}/dapp-request${data ? `#${encodedRequestData}` : ""}`
    );
    return __privateGet(this, _promise);
  }
  close() {
    __privateMethod(this, _StashedPopup_instances, cleanup_fn).call(this);
    __privateGet(this, _popup).close();
  }
}
_popup = new WeakMap();
_version = new WeakMap();
_id = new WeakMap();
_origin = new WeakMap();
_name = new WeakMap();
_chain = new WeakMap();
_promise = new WeakMap();
_resolve = new WeakMap();
_reject = new WeakMap();
_interval = new WeakMap();
_listener = new WeakMap();
_StashedPopup_instances = new WeakSet();
cleanup_fn = function() {
  if (__privateGet(this, _interval)) {
    clearInterval(__privateGet(this, _interval));
    __privateSet(this, _interval, null);
  }
  window.removeEventListener("message", __privateGet(this, _listener));
};
const _StashedHost = class _StashedHost {
  constructor(request) {
    __privateAdd(this, _request);
    if (typeof window === "undefined" || !window.opener) {
      throw new Error(
        "StashedHost can only be used in a window opened through `window.open`. `window.opener` is not available."
      );
    }
    __privateSet(this, _request, request);
  }
  static fromPayload(payload) {
    const { requestId, appUrl, appName, version, ...rest } = payload;
    const request = parse(StashedRequest, {
      version,
      requestId,
      appUrl,
      appName,
      ...rest
    });
    return new _StashedHost(request);
  }
  getRequestData() {
    return __privateGet(this, _request);
  }
  sendMessage(payload) {
    window.opener.postMessage(
      {
        id: __privateGet(this, _request).requestId,
        source: "stashed-channel",
        payload,
        version: __privateGet(this, _request).version
      },
      __privateGet(this, _request).appUrl
    );
  }
  close(payload) {
    if (payload) {
      this.sendMessage(payload);
    }
    window.close();
  }
};
_request = new WeakMap();
let StashedHost = _StashedHost;
export {
  DEFAULT_STASHED_ORIGIN,
  StashedHost,
  StashedPopup,
  StashedRequest,
  StashedResponse
};
